src/Compilers/Core/Portable/Operations/BasicBlock.cs

```diff
@@ -1,152 +1,119 @@
 using System.Linq;
 using Microsoft.CodeAnalysis.PooledObjects;
 
-namespace Microsoft.CodeAnalysis.Operations
+namespace Microsoft.CodeAnalysis.FlowAnalysis
 {
-    /// <summary>
-    /// PROTOTYPE(dataflow): Add documentation
-    /// </summary>
-    public enum BasicBlockKind
+    internal partial class ControlFlowGraphBuilder
     {
-        Entry,
-        Exit,
-        Block
-    }
-
-    /// <summary>
-    /// PROTOTYPE(dataflow): Add documentation
-    /// PROTOTYPE(dataflow): We need to figure out how to split it into a builder and 
-    ///                      a public immutable type.
-    /// </summary>
-    public sealed class BasicBlock
-    {
-        private readonly ImmutableArray<IOperation>.Builder _statements;
-        private readonly ImmutableHashSet<BasicBlock>.Builder _predecessors;
-        internal (IOperation Value, Branch Branch) InternalNext;
-        internal (IOperation Condition, bool JumpIfTrue, Branch Branch) InternalConditional;
-
-        public BasicBlock(BasicBlockKind kind)
-        {
-            Kind = kind;
-            _statements = ImmutableArray.CreateBuilder<IOperation>();
-            _predecessors = ImmutableHashSet.CreateBuilder<BasicBlock>();
-        }
-
-        public BasicBlockKind Kind { get; private set; }
-        public ImmutableArray<IOperation> Statements => _statements.ToImmutable();
-
-        /// <summary>
-        /// PROTOTYPE(dataflow): Tuple is temporary return type, we probably should use special structure instead.
-        /// </summary>
-        public (IOperation Condition, bool JumpIfTrue, Branch Branch) Conditional => InternalConditional;
-
         /// <summary>
-        /// PROTOTYPE(dataflow): During CR there was a suggestion to use different name - "Successor".
+        /// PROTOTYPE(dataflow): Rename this type to "BasicBlockBuilder" and source file to "ControlFlowGraphBuilder.BasicBlockBuilder.cs"
         /// </summary>
-        public (IOperation Value, Branch Branch) Next => InternalNext;
-
-        public ImmutableHashSet<BasicBlock> Predecessors => _predecessors.ToImmutable();
+        internal sealed class BasicBlock
+        {
+            private readonly ArrayBuilder<IOperation> _statementsBuilder;
+            private readonly HashSet<BasicBlock> _predecessorsBuilder;
+            public (IOperation Condition, bool JumpIfTrue, Branch Branch) InternalConditional;
+            public (IOperation Value, Branch Branch) InternalNext;
+            
+            public BasicBlock(BasicBlockKind kind)
+            {
+                Kind = kind;
+                _statementsBuilder = ArrayBuilder<IOperation>.GetInstance();
+                _predecessorsBuilder = new HashSet<BasicBlock>();
+            }
 
-        public int Ordinal { get; internal set; } = -1;
+            public IReadOnlyCollection<IOperation> Statements => _statementsBuilder;
+            public ImmutableArray<IOperation> GetImmutableStatements() => _statementsBuilder.ToImmutableAndFree();
```

Comment:
We should set _statementsBuilder to null when we do this, to make sure we're not actually holding onto a free'd builder.

Reply:
> We should set _statementsBuilder to null when we do this, to make sure we're not actually holding onto a freed builder.

I'll follow  up on that myself

---
In reply to: [192247313](https://github.com/dotnet/roslyn/pull/27254#discussion_r192247313) [](ancestors = 192247313)